src/edu/csus/ecs/pc2/ui/ShadowCompareRunsPane.java

```diff
@@ -415,4 +836,51 @@ private String getCSVString(ShadowJudgementInfo info) {
         return retStr;
     }
     
+    /**
+     * 
+     * 
+     * @author John Clevenger, PC2 Development Team (pc2@ecs.csus.edu)
+     */
+    public class RunListenerImplementation implements IRunListener {
+
+        public void runAdded(RunEvent event) {
+            //we don't care about added runs; we're only interested in when runs get EDITED
+        }
+        
+        public void refreshRuns(RunEvent event) {
+            //we'll refresh when we get an Edited Run Updated notification for the run we requested to be edited
+        }
+
+        public void runChanged(RunEvent event) {
+
+            if (event != null) {
+                
+                //get the run for which we have received a server "changed" notification
+                Run updatedRun = event.getRun();
+                
+                //check if the run update was for the run we requested the server to update
+                if (updatedRun.getElementId().equals(runWeRequestedServerToUpdate.getElementId())) {
```

Comment:
This seems odd, not scalable.   Should the requested run be a Map rather than
a single variable (run) ?

Will there ever be the case where there is more than one resolved run/judgement at a time?

Note: if this is not scalable then a logoff and login would be needed to update the data as a workaround

Reply:
In the current implementation, it is processing each update request separately, sending a message (packet) to the server requesting that the current run be updated, then waiting for the server to send back a response.  Since we typically only have a few (or perhaps a few dozen) runs to be resolved, and they are almost always done either one-at-a-time (as the Shadow and Primary teams decide what resolution is required based on the specific case in question) there's not much chance of a scalability issue; the only things that ever have to be resolved are runs where the Shadow and the Primary disagree (which tend to be small in number and have to be resolved individually or in small groups).

We could certainly try making it more efficient by supporting batch handling of multiple updates -- but again, such batches don't happen very often and tend to be small.  I do not have the time to work on what I consider an unnecessary refinement right now, given the pressure of getting this working prior to the upcoming contests.  Feel free to file a separate Technical Deficiency Issue for this if you think it's that important.

Reply:
No need to optimize at this time.  Since the batches will be small, the need to be scalable is moot.